include/irr/builtin/material_compiler/glsl/common.glsl

```diff
@@ -0,0 +1,1386 @@
+mat2x4 instr_fetchSrcRegs(in instr_t i)
+{
+	uvec3 r = instr_decodeRegisters(i);
+	return instr_fetchSrcRegs(i, r);
+}
+
+void setCurrInteraction(in vec3 N)
+{
+	vec3 campos = irr_glsl_MC_getCamPos();
+	irr_glsl_IsotropicViewSurfaceInteraction interaction = irr_glsl_calcFragmentShaderSurfaceInteraction(campos, WorldPos, N);
+	currInteraction = irr_glsl_calcAnisotropicInteraction(interaction);
+}
+
+//clamping alpha to min MIN_ALPHA because we're using microfacet BxDFs for deltas as well (and NDFs often end up NaN when given alpha=0) because of less deivergence
+//probably temporary solution
+#define MIN_ALPHA 0.0001
+float params_getAlpha(in params_t p)
+{
+	return max(p[PARAMS_ALPHA_U_IX].x,MIN_ALPHA);
+}
+vec3 params_getReflectance(in params_t p)
+{
+	return p[PARAMS_REFLECTANCE_IX];
+}
+vec3 params_getOpacity(in params_t p)
+{
+	return p[PARAMS_OPACITY_IX];
+}
+float params_getAlphaV(in params_t p)
+{
+	return max(p[PARAMS_ALPHA_V_IX].x,MIN_ALPHA);
+}
+vec3 params_getSigmaA(in params_t p)
+{
+	return p[PARAMS_SIGMA_A_IX];
+}
+float params_getBlendWeight(in params_t p)
+{
+	return p[PARAMS_WEIGHT_IX].x;
+}
+vec3 params_getTransmittance(in params_t p)
+{
+	return p[PARAMS_TRANSMITTANCE_IX];
+}
+
+
+bxdf_eval_t instr_execute_cos_eval_DIFFUSE(in instr_t instr, in irr_glsl_LightSample s, in params_t params, in bsdf_data_t data)
+{
+	vec3 refl = params_getReflectance(params);
+	float a = params_getAlpha(params);
+	vec3 diffuse = irr_glsl_oren_nayar_cos_eval(s,currInteraction.isotropic,a*a) * refl;
+	return diffuse;
+}
+
+bxdf_eval_t instr_execute_cos_eval_DIFFTRANS(in instr_t instr, in irr_glsl_LightSample s, in params_t params, in bsdf_data_t data)
+{
+	vec3 tr = params_getTransmittance(params);
+	//transmittance*cos/2pi
+	vec3 c = abs(s.NdotL)*irr_glsl_RECIPROCAL_PI*0.5*tr;
+	return c;
+}
+
+bxdf_eval_t instr_execute_cos_eval_DIELECTRIC(in instr_t instr, in irr_glsl_LightSample s, in float eval)
+{
+	return bxdf_eval_t(eval);
+}
+
+bxdf_eval_t instr_execute_cos_eval_THINDIELECTRIC(in instr_t instr, in irr_glsl_LightSample s, in params_t params, in bsdf_data_t data)
+{
+	/*if (currBSDFParams.isotropic.NdotL>FLT_MIN)
+	{
+		//float au = params_getAlpha(params);
+		//float av = params_getAlphaV(params);
+		vec3 eta = vec3(1.5);
+		vec3 diffuse = irr_glsl_lambertian_cos_eval(currBSDFParams.isotropic,currInteraction.isotropic) * vec3(0.89);
+		diffuse *= irr_glsl_diffuseFresnelCorrectionFactor(eta,eta*eta) * (vec3(1.0)-irr_glsl_fresnel_dielectric(eta, currInteraction.isotropic.NdotV)) * (vec3(1.0)-irr_glsl_fresnel_dielectric(eta, currBSDFParams.isotropic.NdotL));
+		writeReg(REG_DST(regs), diffuse);
+	}
+	else*/ 
+	return bxdf_eval_t(0.0);
+}
+eval_and_pdf_t instr_execute_cos_eval_pdf_THINDIELECTRIC(in instr_t instr, in irr_glsl_LightSample s, in uvec3 regs, in params_t params, in bsdf_data_t data)
+{
+	bxdf_eval_t eval = instr_execute_cos_eval_THINDIELECTRIC(instr, s, params, data);
+	//WARNING 1.0 instead of INF
+	return eval_and_pdf_t(eval, 1.0);
```

Comment:
WHYYY?

Reply:
actually because this function you execute when this BRDF was not the generator... the PDF should be 0.0